src/libcore/slice/sort.rs

```diff
@@ -764,22 +764,27 @@ pub fn partition_at_index<T, F>(v: &mut [T], index: usize, mut is_less: F)
         panic!("partition_at_index index {} greater than length of slice {}", index, v.len());
     }
 
+    let len_minus_one = v.len() - 1;
     if mem::size_of::<T>() == 0 {
         // Sorting has no meaningful behavior on zero-sized types. Do nothing.
-    } else if index == v.len() - 1 {
+    } else if index == len_minus_one {
         // Find max element and place it in the last position of the array
-        for i in 0..(v.len()) {
-            if is_less(&v[index], &v[i]) {
-                v.swap(index, i);
+        let mut max_index = len_minus_one;
+        for i in 0..len_minus_one {
+            if is_less(&v[max_index], &v[i]) {
+                max_index = i;
             }
         }
+        v.swap(max_index, index);
     } else if index == 0 {
         // Find min element and place it in the first position of the array
-        for i in 0..(v.len()) {
-            if is_less(&v[i], &v[index]) {
-                v.swap(index, i);
+        let mut min_index = 0;
+        for i in 1..(v.len()) {
```

Comment:
We can write this with an iterator over the vec, so we don't have to use indexing. iter + enumerate + min_by should work I think

Reply:
I did this, but the code seems a bit uglier since I had to convert the `is_less` into something that returned `Ordering`s. PTAL.

Reply:
Looking, sure, I agree with your taste, especially that we have the unwrap (because the sequence is known to be nonempty, if you have time, put in a comment above the unwrap to say this). Any iterator solution would be better than the old index based one, though.

`.max_by_key` would be a viable replacement to get rid of the Ordering, because you can just return the element reference as the key.

Reply:
I don't follow about `max_by_key`. We need to use `is_less` to evaluate ordering, but it's a binary function, so we can't pass it to `max_by_key`. Could you elaborate?

Reply:
that's my thinko again, that doesn't work, you're right, we have a custom order function. Sorry about that.